### DISCUSSION TITLE
Blazor Server: Hosting Startup Assemblies with loosely coupled dynamically loaded assemblies?

### DISCUSSION THREAD

**Discussion by TheXenocide** [dotnet/runtime#92785] [Q&A] [answered]
NOTE: I considered opening this over on the ASP.NET Core repo, but based on the [related documentation](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/host/platform-specific-configuration?view=aspnetcore-6.0#console-app-without-an-entry-point) it feels like the implementation details I'm curious about may be runtime related.

Our product has a lot of extensibility for our customers and on other platforms has historically provided the opportunity for customers with development capacity to add their own UIs, etc.

Is there some alternative besides registering custom assemblies in the runtime package store we could consider? A lot of our other extension types have a common home. I guess there is some required dependency information generated with the console app but is it possible we could dynamically combine these or otherwise configure our application to allow them to be loaded from the app directory or a custom extension directory?

Our application does provide dynamic compilation of some customizable components during installation of customs in the product, but we'd rather avoid having to dynamically recompile the whole web server application.

Lastly, it would be helpful to understand more about what causes the requirement of project/package references or the runtime package store. We did some prototyping and were able to get the hosting startup assemblies to load dynamically including `IHostingStartup` being executed automatically and the assembly being registered in the router's `AdditionalAssemblies`, but it appears the dependency information may be related to static content/resources, because the pages appear to "render" but are missing all their css/js/etc. Maybe we can overcome this specific issue and use our own integration to load these after all, if this problem is easily solved. I will note that I believe we didn't convert these libraries to console apps with no entry point in them, I think we tried to publish to a directory to see if that would help (but it didn't).

**Comment by maraf** [comment]
> but it appears the dependency information may be related to static content/resources, because the pages appear to "render" but are missing all their css/js/etc.

The problem is `razorclasslib` extracts all static assets during build into target app `wwwroot/_content/{packageid}`. You would either have to copy them to wwwroot manually during extension assembly deployment or invent different mechanism, possibly with embedded resources.

**Reply by TheXenocide** [reply]
So if we can manage the content packaging/deployment, besides the loading of the assemblies (which we have tentatively working) do you anticipate we would run into any other major issues?

**Reply by maraf** [reply, marked as answer]
Nothing that comes to mind, although I'm not expert in this area. I would suggest creating a simple POC.
